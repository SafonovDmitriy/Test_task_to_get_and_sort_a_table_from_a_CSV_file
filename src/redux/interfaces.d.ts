import { PRICING_TABLE_ACTION_TYPE_ENUM } from "./actionTypes/pricingTable";

export interface ITableState {
  data: Array<string[]>;
}
export interface IFetchDataForPricingTable {
  type: PRICING_TABLE_ACTION_TYPE_ENUM;
  payload: Array<string[]>;
}
