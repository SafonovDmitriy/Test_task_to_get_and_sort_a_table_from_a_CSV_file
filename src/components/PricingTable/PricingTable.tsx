import { FC } from "react";
import "./PricingTableStyle.css";

interface IPricingTableProps {
  dataForPricingTable: Array<string[]>;
}

const PricingTable: FC<IPricingTableProps> = ({ dataForPricingTable }) => {
  const checkIfGroup = (tr: string[]): boolean =>
    tr.filter((item) => item).length === 1;

  const renderCheckMark = () => <span className="checkMark">&#10003;</span>;

  return (
    <table className="app_table">
      <tbody>
        {dataForPricingTable.map(
          (
            row,
            rowIdx // It uses a cycle iteration index as a key due to the fact that this massif never mutates
          ) => (
            <tr key={rowIdx} className={checkIfGroup(row) ? "boldText" : ""}>
              {row.map(
                (
                  cell,
                  cellIdx // It uses a cycle iteration index as a key due to the fact that this massif never mutates
                ) => (
                  <td key={`${rowIdx}-${cellIdx}`}>
                    {cell === "X" ? renderCheckMark() : cell}
                  </td>
                )
              )}
            </tr>
          )
        )}
      </tbody>
    </table>
  );
};

export default PricingTable;
