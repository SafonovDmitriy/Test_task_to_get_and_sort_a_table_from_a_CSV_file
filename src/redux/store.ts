import { applyMiddleware, combineReducers, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import thunk from "redux-thunk";
import { pricingTablesReducer } from "./reducers/pricingTablesReducer";

const reducers = combineReducers({
  pricingTables: pricingTablesReducer,
});

const store = createStore(
  reducers,
  composeWithDevTools(applyMiddleware(thunk))
);
export type RootType = ReturnType<typeof reducers>;
export default store;
