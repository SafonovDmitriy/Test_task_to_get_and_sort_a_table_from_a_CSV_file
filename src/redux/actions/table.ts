import { ThunkDispatch } from "redux-thunk";
import { scvSorting } from "../../util";
import svcParser from "../../util/helpers/svcParser";
import { tableActionTypes } from "../actionTypes/table";
import { RootType } from "../store";

export const fetchDataAction = (payload: string[][]) => ({
  type: tableActionTypes.FETCH_DATA,
  payload,
});
export const setLoadingAction = (payload: boolean) => ({
  type: tableActionTypes.SET_LOADING,
  payload,
});
export const fetchData =
  () => async (dispatch: ThunkDispatch<RootType, undefined, any>) => {
    try {
      dispatch(setLoadingAction(true));
      const data = await svcParser("data/data.csv");
      dispatch(fetchDataAction(scvSorting(data)));
    } catch (error) {
      console.log(`error`, error);
    } finally {
      dispatch(setLoadingAction(false));
    }
  };
