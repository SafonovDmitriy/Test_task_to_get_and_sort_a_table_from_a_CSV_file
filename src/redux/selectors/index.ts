import { RootType } from "../store";

export const dataForPricingTablesSelector = (state: RootType) =>
  state.table.data;
