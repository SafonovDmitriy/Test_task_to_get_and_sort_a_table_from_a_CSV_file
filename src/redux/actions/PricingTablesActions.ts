import { Action } from "redux";
import { ThunkDispatch } from "redux-thunk";
import csvParser from "../../helpers/csvParser";
import csvSorting from "../../helpers/csvSorting";
import { pricingTableActionTypes } from "../actionTypes/table";
import { RootType } from "../store";

export const setDataForPricingTablesAction = (payload: string[][]) => ({
  type: pricingTableActionTypes.SET_DATA_FOR_PRICING_TABLES,
  payload,
});

export const fetchDataForPricingTables =
  () => async (dispatch: ThunkDispatch<RootType, undefined, Action>) => {
    try {
      const dataForPricingTables: Array<string[]> = await csvParser(
        "data/data.csv"
      );
      const sortData: Array<string[]> = csvSorting(dataForPricingTables);
      dispatch(setDataForPricingTablesAction(sortData));
    } catch (error) {
      console.log(`error`, error);
    }
  };
