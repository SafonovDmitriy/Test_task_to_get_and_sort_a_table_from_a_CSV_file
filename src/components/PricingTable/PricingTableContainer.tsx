import React, { FC, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { fetchDataForPricingTable } from "../../redux/actions/pricingTablesActions";
import { dataForPricingTableSelector } from "../../redux/selectors";
import PricingTable from "./PricingTable";
//@ts-ignore //False refund in the case of an empty array contradicts the standard FC behavior
const PricingTableContainer: FC = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchDataForPricingTable());
  }, [dispatch]);

  const dataForPricingTable: Array<string[]> = useSelector(
    dataForPricingTableSelector
  );

  return (
    !!dataForPricingTable.length && (
      <PricingTable dataForPricingTable={dataForPricingTable} />
    )
  );
};
export default PricingTableContainer;
