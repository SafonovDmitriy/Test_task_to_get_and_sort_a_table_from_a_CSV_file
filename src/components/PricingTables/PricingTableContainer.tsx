import React, { FC, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { fetchDataForPricingTable } from "../../redux/actions/pricingTablesActions";

import { dataForPricingTableSelector } from "../../redux/selectors";
import PricingTable from "./PricingTable";

const PricingTableContainer: FC = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchDataForPricingTable());
  }, [dispatch]);

  const dataForPricingTable: Array<string[]> = useSelector(
    dataForPricingTableSelector
  );

  return dataForPricingTable.length ? (
    <PricingTable dataForPricingTable={dataForPricingTable} />
  ) : null;
};
export default PricingTableContainer;
