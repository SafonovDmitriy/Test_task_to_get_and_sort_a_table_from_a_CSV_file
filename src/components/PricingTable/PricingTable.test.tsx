import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import * as fetchDataForPricingTable from "../../redux/actions/pricingTablesActions";
import * as dataForPricingTableSelector from "../../redux/selectors";
import PricingTable from "./PricingTable";
import PricingTableContainer from "./PricingTableContainer";

const sortData = [
  ["Names", "Pkg1", "Pkg2", "Pkg3"],
  ["Price (â‚¬)", "5", "50", "100"],
  ["Description", "Pkg1 description", "Pkg2 description", "Pkg3 description"],
  ["Application Intergrations", "", "", ""],
  ["Install custom apps", "X", "-", "-"],
  ["Install new apps", "-", "X", "X"],
  ["Blog", "", "", ""],
  ["Create your own blog", "X", "X", "X"],
  ["Upload Images", "-", "X", "X"],
  ["File Storage", "", "", ""],
  ["Share files with team", "X", "X", "X"],
  ["Upload files", "-", "X", "-"],
  ["Watch Videos", "-", "-", "-"],
  ["Reporting", "", "", ""],
  ["Create your own reports", "X", "X", "X"],
  ["Detailed Reports", "-", "X", "X"],
  ["Weekly Data Reports", "-", "X", "X"],
];

describe("Pricing Table Presentation Component", () => {
  it("render Pricing Presentation Table component", () => {
    const pricingTable = renderer.create(
      <PricingTable dataForPricingTable={sortData} />
    );
    expect(pricingTable.toJSON()).toMatchSnapshot();
  });
});

describe("Pricing Table Container Component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const middlewares = [thunk];
  const initialStore = { pricingTable: { data: [] } };
  const mockStore = configureMockStore(middlewares);
  const store = mockStore(initialStore);

  it("Pricing Table component with sortData", async () => {
    jest
      .spyOn(dataForPricingTableSelector, "dataForPricingTableSelector")
      .mockReturnValue(sortData);
    const pricingTable = renderer.create(
      <Provider store={store}>
        <PricingTableContainer />
      </Provider>
    );

    expect(pricingTable.toJSON()).not.toBe(null);
  });

  it("Pricing Table component with empty data", async () => {
    jest
      .spyOn(dataForPricingTableSelector, "dataForPricingTableSelector")
      .mockReturnValue([]);

    const fetchDataForPricingTableSpy = jest.spyOn(
      fetchDataForPricingTable,
      "fetchDataForPricingTable"
    );

    const pricingTable = renderer.create(
      <Provider store={store}>
        <PricingTableContainer />
      </Provider>
    );

    expect(pricingTable.toJSON()).toBe(null);
    expect(fetchDataForPricingTableSpy).toBeCalled();
  });
});
