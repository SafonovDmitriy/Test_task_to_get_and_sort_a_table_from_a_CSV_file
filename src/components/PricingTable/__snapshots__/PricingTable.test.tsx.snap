// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pricing Table Presentation Component render Pricing Presentation Table component 1`] = `
<table
  className="app_table"
>
  <tbody>
    <tr
      className=""
    >
      <td>
        Names
      </td>
      <td>
        Pkg1
      </td>
      <td>
        Pkg2
      </td>
      <td>
        Pkg3
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Price (€)
      </td>
      <td>
        5
      </td>
      <td>
        50
      </td>
      <td>
        100
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Description
      </td>
      <td>
        Pkg1 description
      </td>
      <td>
        Pkg2 description
      </td>
      <td>
        Pkg3 description
      </td>
    </tr>
    <tr
      className="boldText"
    >
      <td>
        Application Intergrations
      </td>
      <td>
        
      </td>
      <td>
        
      </td>
      <td>
        
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Install custom apps
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        -
      </td>
      <td>
        -
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Install new apps
      </td>
      <td>
        -
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
    </tr>
    <tr
      className="boldText"
    >
      <td>
        Blog
      </td>
      <td>
        
      </td>
      <td>
        
      </td>
      <td>
        
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Create your own blog
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Upload Images
      </td>
      <td>
        -
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
    </tr>
    <tr
      className="boldText"
    >
      <td>
        File Storage
      </td>
      <td>
        
      </td>
      <td>
        
      </td>
      <td>
        
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Share files with team
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Upload files
      </td>
      <td>
        -
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        -
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Watch Videos
      </td>
      <td>
        -
      </td>
      <td>
        -
      </td>
      <td>
        -
      </td>
    </tr>
    <tr
      className="boldText"
    >
      <td>
        Reporting
      </td>
      <td>
        
      </td>
      <td>
        
      </td>
      <td>
        
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Create your own reports
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Detailed Reports
      </td>
      <td>
        -
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
    </tr>
    <tr
      className=""
    >
      <td>
        Weekly Data Reports
      </td>
      <td>
        -
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
      <td>
        <span
          className="checkMark"
        >
          ✓
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;
